Replace api.Rule with types.PolicyEntry

TheBeeZee opened the pull request:
Refactor policy engine to use PolicyEntry as the internal representation of policies, as described in CFP-39646.

Fixes: #40015

```release-note
Refactor policy engine to use PolicyEntry as the internal representation of policies, as described in CFP-39646.
```

TheBeeZee committed:
Replace api.Rule with types.PolicyEntry

Signed-off-by: Blaz Zupan <blaz@google.com>

nimishamehta5 commented on pkg/policy/rule.go:
```diff
@@ -450,27 +432,6 @@ func mergeIngress(policyCtx PolicyContext, fromEndpoints api.EndpointSelectorSli
 		return found, err
 	}
 
-	err = icmp.Iterate(func(r api.Ports) error {
```
What was the background for special handling of ICMP rules, and does it affect anything if we handle them generically?

joestringer replied:
(You could consider finding the PR for that feature or running `git blame` on the corresponding paths on the `main` branch and see if they left context in the commit messages, or CC the author to see if they have context)

TheBeeZee replied:
As far as I can see from the code, it was just a very verbose (and inefficient) implementation. The only difference between ICMP and the rest is the log message, which specifically says "ICMP code" instead of "port", other than that the logic was same. I could make the log message in the new implementation the same by adding some additional logic, but I'm not sure it is worth the hassle. Feedback appreciated.

squeed commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,77 @@
+	EndpointSelector api.EndpointSelector
+
+	// L3 specifies the source/destination endpoints or all endpoints if empty
+	L3 EndpointSelectorInterfaceSlice
```
I'm so happy this works! So nice and clean.

TheBeeZee replied:
For now I only have EndpointSelector and FQDNSelector as implementations of EndpointSelectorInterfaceSlice. I'm thinking of renaming this interface to something shorter, suggestions appreciated.

I think we could also make CIDRSlice implement EndpointSelectorInterfaceSlice, but I need to take a closer look as my last attempt before going on vacation has failed.

squeed commented on pkg/labels/labels.go:
```diff
@@ -805,6 +805,13 @@ func ParseLabel(str string) Label {
+// ParseK8sLabel returns the label representation of the given string. The str should be
+// in the form of Source.Key=Value or Source.Key if Value is empty. It also parses short
+// forms, for example: $host will be Label{Key: "host", Source: "reserved", Value: ""}.
+func ParseK8sLabel(str string) Label {
```
Right, this is used only for the internal hacky representation of labels to squeeze past the k8s matching library; is there any way we can isolate it?

TheBeeZee replied:
Ah yes, you are right, I forgot to clean this up. I've updated the PR so it now passes CIDRSlice and CIDRRuleSlice as EndpointSelectorSliceInterface, which makes ParseK8sLabel obsolete.

bimmlerd commented on pkg/k8s/apis/cilium.io/utils/repository.go:
```diff
@@ -0,0 +1,141 @@
+	"github.com/cilium/cilium/pkg/policy/types"
+)
+
+// RulesToPolicyEntries converts a slice of api.Rules to PolicyEntries
```
nit: the comment tells me exactly as much as the type signature of the function - would be more interesting to have a comment here which explains (briefly) _why_ we want to have this in the first place. Something like "RulestoPolicyEntries converts the external rule type to the internal policy entry representation" which already gives a bit more context

TheBeeZee replied:
I agree that the previous description was not helpful and have changed it to what you suggest.

bimmlerd commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+// PolicyEntry specifies the L3/L4 details of a single policy rule
+type PolicyEntry struct {
```
there's a lot of bools in this struct, which makes it quite a bit bigger than necessary. Are all combinations of these bools valid? Can we turn the bools into an enum or a bitfield? The enum would be able to only capture valid combinations; the bitfield would be somewhat more ergonomic if all combinations are valid. Both save memory.

TheBeeZee replied:
Yes, all the permutations of the booleans are valid.

Is memory use really a concern? While converting the booleans to a bitfield is certainly an option, it would save a grand total of 3 bytes per policy entry. Even when there are 100000 policy entries present, this would only save about 300k of memory, which in 2025 seems insignificant.

Converting to a bitfield would mean adding getters/setters for each bit, which will make the policy struct needleslly verbose (IMHO).

bowei replied:
I think this is one that can be easily done after the fact with measurement if it turns out to be the major memory consumer. The number of user-sourced dynamically allocated strings in the Go-based code base probably would have a bigger impact on memory consumption than bit-set packing?

bimmlerd replied:
> While converting the booleans to a bitfield is certainly an option, it would save a grand total of 3 bytes per policy entry.

Thought a bit more about this - with alignment and all I don't think we'd by saving anything :sweat_smile:

Let's compromise on rearranging the struct, since that's a lot less intrusive. This changes the size from 152 to 144 bytes, and moves all the pointers to the top so that the GC has to scan less - courtesy of the [fieldalignment](https://cs.opensource.google/go/x/tools/+/refs/tags/v0.35.0:go/analysis/passes/fieldalignment/fieldalignment.go) tool.

```go
type PolicyEntry struct {
	Authentication   *api.Authentication
	Log              api.LogConfig
	EndpointSelector api.EndpointSelector
	L3               EndpointSelectorInterfaceSlice
	L4               api.PortRules
	Labels           labels.LabelArray
	DefaultDeny      bool
	Deny             bool
	Ingress          bool
	Node             bool
}
```

TheBeeZee replied:
Sure, sounds reasonable. Updated.

bimmlerd reviewed (changes requested):
Generally looks really good to me- the policy entry type imo should be modeled better with less bools though.

MrFreezeex reviewed (approved):
ack for the small clustermesh change, thanks!

bimmlerd reviewed (approved):
LGTM - please do wait for @squeed's review though; once he's back.

squeed commented on pkg/k8s/apis/cilium.io/utils/repository.go:
```diff
@@ -0,0 +1,142 @@
+		es, node := getSelector(rule)
+
+		for _, iRule := range rule.Ingress {
+			defaultDeny := rule.EnableDefaultDeny.Ingress != nil && *rule.EnableDefaultDeny.Ingress
```
Not quite sure this is correct - `defaultDeny` is always true unless there is an explicit (non-nil) value.

(this applies to the whole function)

TheBeeZee replied:
Indeed, the logic was inverted. I have modified it so that it is enabled if not set.

squeed commented on pkg/k8s/apis/cilium.io/utils/repository.go:
```diff
@@ -0,0 +1,142 @@
+			l3 := types.ToEndpointSelectorInterfaceSlice(iRule.GetSourceEndpointSelectorsWithRequirements(req))
+			l3 = append(l3, types.ToEndpointSelectorInterfaceSlice(iRule.FromCIDR)...)
+			l3 = append(l3, types.ToEndpointSelectorInterfaceSlice(iRule.FromCIDRSet)...)
+			l4 := append(iRule.ToPorts, icmpRules(iRule.ICMPs)...)
```
I forget the Go slice semantics here. Is there a chance we accidentally mutate `iRule`?

TheBeeZee replied:
Good catch. Append will modify the first slice if the array size is not large enough, which in this case it wouldn't be. I have modified the code so that it now creates a new copy.

squeed commented on pkg/policy/utils/parserules.go:
```diff
@@ -0,0 +1,142 @@
+)
+
+// RulestoPolicyEntries converts the external rule type to the internal policy entry representation.
+func RulesToPolicyEntries(rules api.Rules) types.PolicyEntries {
```
this whole function is super clean. I really like how it lets us simplify so much of the codebase.

squeed commented on pkg/k8s/network_policy.go:
```diff
@@ -140,35 +141,37 @@ func hasV1PolicyType(pTypes []slim_networkingv1.PolicyType, typ slim_networkingv
 		if len(iRule.From) > 0 {
 			for _, rule := range iRule.From {
-				ingress := api.IngressRule{}
+				ingress := &types.PolicyEntry{Ingress: true}
```
should set `DefaultDeny: true` here?

TheBeeZee replied:
No need, because we set it for all rules in the for loop at the end of the function.

squeed commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1804,7 +1806,7 @@ func PolicyDefaultLocalClusterInspect(ctx context.Context, k8sClient *k8s.Client
 		}
 
-		res.NetworkPolicies[client.ObjectKeyFromObject(&np).String()] = !slices.EqualFunc(rulesAny, rulesLocal, func(a, b *api.Rule) bool { return a.DeepEqual(b) })
+		res.NetworkPolicies[client.ObjectKeyFromObject(&np).String()] = !slices.EqualFunc(rulesAny, rulesLocal, func(a, b *types.PolicyEntry) bool { return reflect.DeepEqual(a, b) })
```
should we add DeepEqual to PolicyEntry?

TheBeeZee replied:
Done.

squeed commented on daemon/cmd/1_next_fail/ep_config.h:
```diff
@@ -0,0 +1,20 @@
+/*
```
this seems like an error.

TheBeeZee replied:
Removed.

jrajahalme commented on daemon/cmd/1_next_fail/ep_config.h:
```diff
@@ -0,0 +1,20 @@
+ * - unspec:bar
+ * - unspec:QA
+ */
+
```
Adding directory `daemon/cmd/1_next_fail` is likely a mistake?

TheBeeZee replied:
Fixed.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -450,10 +451,8 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
-				for _, er := range rule.Egress {
-					_ = er.ToPorts.Iterate(removeL7DNSRules)
-				}
+			d.policy.Iterate(func(rule *policytypes.PolicyEntry) {
+				_ = rule.L4.Iterate(removeL7DNSRules)
```
Previously this only iterated egress rules, could this be problematic?

TheBeeZee replied:
Not really a problem. FQDN rules only exist in egress policies. Although this will now iterate over all policies (as they are one flat list), it will only ever remove FQDN rules from egress rules, as they can't possibly exist elsewhere (there's no FQDN ingress rules).

joestringer commented on pkg/policy/k8s/cilium_network_policy_test.go:
```diff
@@ -77,12 +77,12 @@ func Test_GH33432(t *testing.T) {
-	assert.Equal(t, []api.PortProtocol{{Port: "80", Protocol: api.ProtoTCP}}, rules[0].Egress[0].ToPorts[0].Ports)
-	assert.Nil(t, rules[0].Egress[0].EgressCommonRule.ToEndpoints)
+	assert.False(t, rules[0].Ingress)
+	assert.Len(t, rules[0].L3, 2)
```
How did `rules[0].L3` grow to two elements?

TheBeeZee replied:
I was storing CIDRs as both CIDR/CIDRRule and converted into an EndpointSelector. It turns out that this is pretty easy to sort out, so I've commited a change that does that.

joestringer replied:
This doesn't appear to be resolved. If the number of internal rules is an implementation detail that's in flux then maybe we should just drop the assertion of the number of rules.

TheBeeZee replied:
Fixed by latest commit.

joestringer commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,113 @@
+	Log api.LogConfig
+
+	// EndpointSelector specifies the endpoint that this rule applies to
+	EndpointSelector api.EndpointSelector
```
nit, but we can defer to a separate PR: I think clarity could benefit from referring to this as the Subject.
```suggestion
	// Subject specifies the endpoint that this rule applies to
	Subject api.EndpointSelector
```

TheBeeZee replied:
I like this suggestion. I have added a new commit to this PR that changes the name from EndpointSelector to Subject.

joestringer commented on pkg/policy/types/zz_generated.deepequal.go:
How was this generated, is it already taken care of in the standard make targets or did you run some specific commands to create it?

If this is not part of the `make generate-*` targets, please extend those targets to ensure we always generate these files as part of the build.

TheBeeZee replied:
It is generated using deepequal-gen, which is executed as part of `make generate-k8s-api`, based on comments above the PolicyEntry struct (+deepequal-gen=true).

joestringer commented on pkg/policy/k8s/service_test.go:
```diff
@@ -381,50 +376,47 @@ func TestPolicyWatcher_updateToServicesPolicies(t *testing.T) {
-	assert.Equal(t,
-		sortCIDRSet(byNameRules[1].Egress[0].ToCIDRSet),
-		sortCIDRSet(oldByNameRules[1].Egress[0].ToCIDRSet))
+	assert.Equal(t, byNameRules, oldByNameRules)
```
Is sorting not relevant here any more?

TheBeeZee replied:
Very good point, I've seen some flakiness from this test, so it is indeed needed. I have refactored the test so that it can still use sortCIDRSet.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -964,10 +964,12 @@ func createL4Filter(policyCtx PolicyContext, peerEndpoints api.EndpointSelectorS
 	}
 
-	if peerEndpoints.SelectsAllEndpoints() {
+	es := api.EndpointSelectorSlice(types.FromEndpointSelectorInterfaceSlice[api.EndpointSelector](peerEndpoints))
+	if es.SelectsAllEndpoints() {
 		l4.wildcard = l4.cacheIdentitySelector(api.WildcardEndpointSelector, origin.stringLabels(), selectorCache)
 	} else {
-		l4.cacheIdentitySelectors(peerEndpoints, origin, selectorCache)
+		l4.cacheIdentitySelectors(es, origin, selectorCache)
+		fqdns := types.FromEndpointSelectorInterfaceSlice[api.FQDNSelector](peerEndpoints)
 		l4.cacheFQDNSelectors(fqdns, origin, selectorCache)
 	}
```
This seems to iterate and handle all endpoint selectors in `peerEndpoints` that have types `api.EndpointSelector` and `api.FQDNSelector`. Grepping for `FromEndpointSelectorInterfaceSlice`, I also see that function being used for `api.CIDRRule` and `api.CIDR`. Does this mean that creating an L4 filter could omit CIDR-based peer selectors? Does this logic need to be extended for other types as well? If yes, how do we avoid silently dropping valid endpoint selector types with this new generic structure?

joestringer replied:
Short answer from [the other thread](https://github.com/cilium/cilium/pull/40213#discussion_r2373563729) is that the CIDR-based selectors are being converted into `api.EndpointSelector` as they're put into `peerEndpoints` so the `es` unboxing here should include CIDR selectors as well.

(If the PR broke CIDR selection, I would think that at least some of the tests should have failed so I don't think there's much chance of regression here).

joestringer commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,113 @@
+// EndpointSelectorInterface is a generic representation of an endpoint selector.
+// It can be converted to an EndpointSelector or FQDNSelector.
+type EndpointSelectorInterface interface {
+	IsEndpointSelectorInterface()
```
Should this just use an actual real function that these types satisfy, such as `GetAsEndpointSelectors()`?
```suggestion
	GetAsEndpointSelectors()
```

TheBeeZee replied:
GetAsEndpointSelector() is a method of the slice types (i.e. EndpointSelectorSlice, FQDNSlice, CIDRSlice), while this interface is supposed to represent an individual endpoint selector. The only common method they have is String(), which seems pretty generic.

joestringer replied:
I poked around a little in the usage of `FromPeerSelectorSlice()` and the conclusion I came to is just that it's a generic box to hide implementation details of various selectors, but different users are unpacking and doing different implementation-specific things with the resulting selectors so it's not as simple as just picking a good common function or for example adding a `GetAsEndpointSelectors()` method on all the relevant types.

I'm left wondering whether the generic could be somehow improved or taken further as this pattern "feels" a bit off, but maybe I'm biting off more than is worth chewing right now. This is something we can leave for a future iteration.

joestringer commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,113 @@
+
+// EndpointSelectorInterface is a generic representation of an endpoint selector.
+// It can be converted to an EndpointSelector or FQDNSelector.
+type EndpointSelectorInterface interface {
```
It seems like this is always used to select peers, maybe it would be simpler to refer to this as:
```suggestion
// PeerSelector is a generic representation of a selector for network peers in a policy rule.
type PeerSelector interface {
```

joestringer replied:
(I also dropped the `It can be converted ...` since that seems like it's already an out-of-date or not comprehensive comment)

TheBeeZee replied:
Thank you for this suggestion. As you can see, I was struggling with naming. I like PeerSelector, it is short and concise. I have added a commit that changes the name.

joestringer commented on pkg/k8s/network_policy.go:
```diff
@@ -140,35 +141,37 @@ func hasV1PolicyType(pTypes []slim_networkingv1.PolicyType, typ slim_networkingv
-					ingress.FromCIDRSet = append(ingress.FromCIDRSet, ipBlockToCIDRRule(rule.IPBlock))
+					cidrRules := api.CIDRRuleSlice{ipBlockToCIDRRule(rule.IPBlock)}
+					ingress.L3 = append(ingress.L3, types.ToEndpointSelectorInterfaceSlice(cidrRules.GetAsEndpointSelectors())...)
+					ingress.L3 = append(ingress.L3, types.ToEndpointSelectorInterfaceSlice(cidrRules)...)
```
Why is this adding `cidrRules` or a derivative of `cidrRules` to `ingress.L3` twice?

joestringer replied:
(Same question for egress)

joestringer replied:
I saw a similar pattern in the tests, so I wonder if the tests were made to duplicate some of these and that then motivated duplicating them in the production code too?

joestringer replied:
Coming back to this later I wonder if this is a confusion when implementing k8s netpol. Upstream kubernetes netpol only has `IPBlock`. On the other hand, CNP has two fields - `fromCIDR` and `fromCIDRSet`. So this case looks duplicate, whereas for CNP there is a legitimate need to copy two CIDR-related fields.

TheBeeZee replied:
This was removed in my last commit.

joestringer replied:
Hmm. It seems like the second line here might be being used to populate ipcache, while maybe the first line is actually in the end used to perform the selections in the policy engine. This relates to the [discussion in the other thread above](https://github.com/cilium/cilium/pull/40213#discussion_r2356296880).

Right now my conclusion for KNP is that the behavior is correct but the code is confusing. Maybe if CIDR selectors were first class selectors then they could also have a convenient helper to translate them into a slice of `netip.Prefix` for use in the ipcache plumbing code. That said, if the logic is correct then maybe we can punt that out to a separate PR to avoid this one growing too much bigger.

(CNP is another path so I'll look at that separately).

joestringer commented on pkg/k8s/network_policy.go:
```diff
@@ -257,23 +270,19 @@ func ParseNetworkPolicy(logger *slog.Logger, clusterName string, np *slim_networ
-		WithIngressRules(ingresses).
-		WithEgressRules(egresses)
-
-	if err := rule.Sanitize(); err != nil {
-		return nil, err
```
What happened to sanitization?

It looks like previously this will have translated everything into the internal format, then sanitized the rules to ensure they follow an expected form; if not, reject the whole policy. Now that the internal format is different, should there be an earlier phase to validate the rules before translating to the internal format?

TheBeeZee replied:
Sanitization that applied to api.Rule doesn't make sense here. Some examples:

1. It checks that either NodeSelector or EndpointSelector is populated. We only have an EndpointSelector (as `Subject`), so this check is irrelevant.
2. It checks that either Ingress/Egress/IngressDeny/EgressDeny are populated. Again, this does not apply, as we always populate at least a default deny rule in ingress/egress direction.
3. It checks default deny policy, which only makes sense in the context of Cilium network policies. We always set the default deny policy in this function.
4. It performs checks on L7 policies, which do not exist in the context of k8s network policies.
5. It validates label selectors. In this function, the label selector is synthesized, so it is guaranteed to be correct (the check was redundant).

Effectively, it is the responsibility of this function to produce valid PolicyEntries and to sanitize all inputs as it is iterating through them.

joestringer replied:
Agree in general that most of the sanitization is regarding CNP-specific features, thanks for looking into it.

> It validates label selectors. In this function, the label selector is synthesized, so it is guaranteed to be correct (the check was redundant).

This was the most interesting one, will think out loud here:
`parseNetworkPolicyPeer()` is generating the selector with a call to `api.NewESFromK8sLabelSelector()` which embeds the k8s source properly. This will have been happening with existing code too, and would prepare the selector with the appropriate source. This function calls `NewESFromK8sLabelSelectorWithExtender()` with the caller's source prefix, ie k8s prefix. The now-deleted `EndpointSelector.Sanitize()` would also directly call `NewESFromK8sLabelSelectorWithExtender()` but with the previously-parsed & prepared selector that has the source embedded, so presumably that was effectively a noop (it's just that it was conveniently sharing the sanitization with the CNP codepaths). 

So yeah I come to the same conclusion that all the sanitization stuff is for handling interesting CNP semantics that don't apply for KNP. Beauty of simplicity for the smaller API, I guess.

joestringer replied:
The only remaining part I'm not sure about is the call to `SetAggregatedSelectors` which is effectively preparing a cached selection for a rule. This was an optimization we rolled out at some point due to selectors being expensive. It doesn't directly translate with the new PolicyEntry structure but I haven't got my head around whether that would have a meaningful performance impact at this stage. Maybe we can just take a note to see what kind of benchmarks we can use to evaluate policy calculation speed and aim to evaluate that prior to the next release (+ roll out mitigations if necessary).

joestringer replied:
Looking a little closer at `SetAggregatedSelectors`, I think another subsequent follow-up that would be worth cleaning up would be to delete all of that logic. If the API format is no longer the internal format then we don't need those optimizations any more.

At the same time we could dig through git history to when this cache was introduced and maybe there'll be some commentary about the performance impact and how it was assessed.

TheBeeZee replied:
Yes, I did just that in my latest commit, check it out. SetAggregatedSelectors is gone completely.

joestringer commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,113 @@
+			if v2, ok := other.L3[i].(api.EndpointSelector); !ok || !v1.DeepEqual(&v2) {
+				return false
+			}
+		case api.CIDR:
+			if v2, ok := other.L3[i].(api.CIDR); !ok || v1 != v2 {
+				return false
+			}
+		case api.CIDRRule:
+			if v2, ok := other.L3[i].(api.CIDRRule); !ok || !v1.DeepEqual(&v2) {
+				return false
+			}
```
Why is there both CIDR and CIDRRule?

joestringer replied:
Answer: This corresponds to `fromCIDR` vs `fromCIDRSet` in CNPs (or equivalent egress CIDR peer match types).

joestringer commented on pkg/policy/rules.go:
```diff
@@ -18,27 +15,9 @@ func (rules ruleSlice) resolveL4IngressPolicy(policyCtx PolicyContext) (L4Policy
-
-	for _, r := range rules {
-		err := r.resolveIngressPolicy(policyCtx, &state, result, requirements, requirementsDeny)
+		err := r.resolveIngressPolicy(policyCtx, &state, result)
```
If I follow right the intent here is to move requirements calculation out from the resolution / policy calculation step and instead move it all the way back to initial parsing so that the requirements are integrated into the ruleset from the start. Given that k8s has had requirements for quite some time now it feels like this could naturally fit into the translation as a completely separate logical change.

I wonder whether it would be possible to split this logical change out into a first commit or even separate PR. Requirements are tricky to get our head around at the best of times, and it would be really easy to introduce a regression in the middle of this 2K+ LoC PR because there's too many other changes going on.

squeed replied:
IIRC the tricky bit about Requirements is that they are merged from all policies for a given endpoint. That is to say, if policy A has Requirements, then policy B implicitly has those requirements as well. And this is a **per identity** state.

This sort of "spooky action at a distance" is why I would like to remove Requirements, but I'm not super comfortable doing deprecating a stable policy feature -- and even less so one that would make policies more liberal.

joestringer replied:
Ah that probably means this PR breaks requirements right? Because the requirements from one policy are only being folded into rules of the same policy and not applied to all other policies?

Or is there some later stage where the requirements are merged globally?

TheBeeZee replied:
I have fixed the FromRequires/ToRequires support (see latest commit in this PR). I think it would be easier to read if I squash all the commits into the main one, but for now I'll leave it separate so you can take a look.

joestringer replied:
I glanced over that change and it seems relatively straightforward (though I don't understand the deepequal gen piece or how it relates). In this case yes I think it makes sense to fold back into the original commit. That should serve to also reduce the size and scope of that original commit, which is a good thing.

TheBeeZee replied:
This is fixed now, we iterate over requirements in mergeIngress / mergeEgress as before.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -450,27 +432,6 @@ func mergeIngress(policyCtx PolicyContext, fromEndpoints api.EndpointSelectorSli
 		return found, err
 	}
 
-	err = icmp.Iterate(func(r api.Ports) error {
```
Moving ICMP iteration steps into L4 would be another candidate for a separate change, again it's hard to track this across the size of this commit/PR.

TheBeeZee replied:
Done.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -484,19 +445,19 @@ func (r *rule) resolveIngressPolicy(
-		cnt, err := mergeIngress(policyCtx, fromEndpoints, ingressRule.Authentication, ingressRule.ToPorts, ingressRule.ICMPs, result)
+
+	if !r.Deny {
+		cnt, err := mergeIngress(policyCtx, r.L3, r.Authentication, r.L4, result)
```
Where does this iteration / merging occur to ensure merging of different rules from the API layer if this step is now doing a boolean check?

TheBeeZee replied:
It happens in pkg/k8s/apis/cilium.io/utils/repository.go.

joestringer replied:
Ah right I see, because the new internal representation has one flat layer of policy entries rather than having a two tier layer of rules (really policies?) which each have rules inside. 👌

joestringer commented on pkg/metrics/features/policy.go:
Where did this logic end up so we can review / compare with existing logic?

TheBeeZee replied:
To obtain metrics, the existing code has to awkwardly iterate over all the sub-structures of api.Rule. Because we now parse all api.Rule into a single flat slice of PolicyEntries, the logic is greatly simplified. The conversion from api.Rules to types.PolicyEntries happens in k8s/apis/cilium.io/utils/repository.go.

joestringer replied:
It is a lot tidier, nice.

Minor nit I would suggest for future is to provide this kind of context in the commit message. The reason I asked this question multiple times was because there was a key design change here in flattening the internal representation, and that kind of high level design decision + rationale can be hard to extract from the code.  Even if this may have been discussed in the CFP, future readers going back through this commit or bisecting a bug to this commit aren't going to have that reference link handy. What they will have is the git logs / commits and code.

joestringer reviewed (commented):
Thanks for taking this on, I think this will pay off over time in generalizing the internal representation. 

I left a bunch of comments all over and I still didn't get to the bottom. Unfortunately the size of the change is making this just difficult to work through. It's hard to tell where things are moving and whether important assumptions are holding. I know this is a pretty deep change in the core of policy so it's hard to really split up, but if there is any way to do so then I think that'd likely speed up some of the review and give reviewers a bit more confidence. Either way I figure it's better to push out this review now even if not complete to give something to discuss rather than drag it out further.

joestringer commented on pkg/k8s/apis/cilium.io/utils/repository.go:
```diff
@@ -0,0 +1,169 @@
+
+			req := convertToLabelSelectorRequirementSlice(iRule.FromRequires)
+			endpoints := iRule.GetSourceEndpointSelectorsWithRequirements(req)
+			l3 := mergeEndpointSelectors(endpoints, iRule.FromCIDR, iRule.FromCIDRSet, nil)
```
I think there's some duplication happening here :thinking: 

*`(i *IngressCommonRule) SetAggregatedSelectors()`*
```
        res := make(EndpointSelectorSlice, 0, len(i.FromEntities)+len(i.FromCIDR)+len(i.FromCIDRSet))                                                                                                                             
        res = append(res, i.FromEntities.GetAsEndpointSelectors()...)                                                                                                                                                             
        res = append(res, i.FromCIDR.GetAsEndpointSelectors()...)                                                                                                                                                                 
        res = append(res, i.FromCIDRSet.GetAsEndpointSelectors()...)                                                                                                                                                              
                                                                                                                                                                                                                                  
        i.aggregatedSelectors = res
```
... *`(i *IngressCommonRule) GetSourceEndpointSelectorsWithRequirements(...)`*
```
        return append(res, i.aggregatedSelectors...)
```
... *`mergeEndpointSelectors(...)`*
```
         l3 := make(types.PeerSelectorSlice, 0, len(endpoints)+len(cidrSlice)+len(cidrRuleSlice)+len(fqdns))                                                                                                                     
         l3 = append(l3, types.ToPeerSelectorSlice(endpoints)...)                                                                                                                                                                
         l3 = append(l3, types.ToPeerSelectorSlice(cidrSlice)...)                                                                                                                                                                
         l3 = append(l3, types.ToPeerSelectorSlice(cidrRuleSlice)...)                                                                                                                                                            
         l3 = append(l3, types.ToPeerSelectorSlice(fqdns)...)     ■ in call to types.ToPeerSelectorSlice, T (type api.FQDNSelector) does not satisfy "github.com/cilium/cilium/pkg/policy/types".PeerSelector (missing method IsP
         return l3
```

joestringer replied:
I'm not exactly sure I understand the implications of this, and I guess it'll be resolved with the commit you have pending to rip out aggregated selectors. If this isn't breaking tests then maybe we just punt on this for now since the same problem will be addressed soon enough anyway.

TheBeeZee replied:
Yes, I have a follow-up commit (it's in this PR, take a look) to remove this duplication.

TheBeeZee replied:
Reopening this, as the latest changes broke some unit tests.

TheBeeZee replied:
All unit tests are fixed. I believe this fixes the issue with requirements.

TheBeeZee committed:
Consolidate handling of ICMP and other port protocols

Signed-off-by: Blaz Zupan <blaz@google.com>

TheBeeZee committed:
Mark Test*PolicyEtcd tests as privileged

Signed-off-by: Blaz Zupan <blaz@google.com>

joestringer commented on pkg/policy/rule.go:
```diff
@@ -484,19 +442,19 @@ func (r *rule) resolveIngressPolicy(
+		return nil
 	}
-	for _, ingressRule := range r.Ingress {
-		fromEndpoints := ingressRule.GetSourceEndpointSelectorsWithRequirements(requirements)
```
The removal of requirements here means that the comment above the function should be updated to reflect the change.

TheBeeZee replied:
Requirements are back (in latest commit), so comment stays as-is.

joestringer commented on pkg/metrics/features/policy.go:
```diff
@@ -180,63 +147,27 @@ func ruleTypePortRules(rf *RuleFeatures, portRules api.PortRules) {
+	rf.MutualAuth = r.Authentication != nil
+	rf.NonDefaultDeny = r.DefaultDeny
+	rf.ToFQDNs = len(types.FromPeerSelectorSlice[api.FQDNSelector](r.L3)) > 0
+	rf.L3 = len(r.L3) > 0 && !rf.ToFQDNs
```
This doesn't seem right, but the bug seems to be present in the tree already. FQDNSelectors are an L3 selector so it's weird that feature reporting doesn't treat it as such.
```suggestion
	rf.L3 = len(r.L3) > 0
```

joestringer replied:
(Maybe worth deferring this fix to a future PR so we can pull in the relevant authors to understand whether this bugfix needs backporting)

TheBeeZee replied:
I'll mark this resolved as this is existing behavior. Let's revisit in a future PR.

joestringer commented on pkg/policy/api/selector.go:
```diff
@@ -386,16 +388,6 @@ func (n *EndpointSelector) Sanitize() error {
-func (s EndpointSelectorSlice) Len() int      { return len(s) }
-func (s EndpointSelectorSlice) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
-
-func (s EndpointSelectorSlice) Less(i, j int) bool {
-	strI := s[i].LabelSelectorString()
-	strJ := s[j].LabelSelectorString()
-
-	return strings.Compare(strI, strJ) < 0
-}
-
```
I'm surprised we can delete these, don't we need to sort the slice for reliable output somewhere in testing code?

TheBeeZee replied:
I put it back, just to reduce the amount of unneccessary diffs.

joestringer commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,112 @@
+		return false
+	}
+	for i := range in.L3 {
+		switch v1 := in.L3[i].(type) {
```
Should `FQDNSelector` be part of the deepequal method here?

TheBeeZee replied:
Good catch. Added.

joestringer replied:
Incidentally, this makes me wonder whether part of the `L3` interface should include DeepEqual method then this code doesn't need to unwrap / handle it (and potentially miss certain selector types like this)

joestringer commented on pkg/policy/utils/parserules.go:
Why is this in the APIs directory? It's about taking an external API into the internal format. It should be part of the internal pkg/policy or pkg/policy/utils if necessary.

(Reason I ask is that this is somehow not owned by @cilium/sig-policy , but it should be).

TheBeeZee replied:
It naturally belongs in either pkg/policy or pkg/policy/api, but I can't put it in either of those because it would cause a circular dependency. pkg/policy/utils seems feasible though.

TheBeeZee replied:
I have moved this to pkg/policy/utils.

joestringer commented on pkg/policy/repository_deny_test.go:
```diff
@@ -495,13 +472,7 @@ func TestL3DependentL4EgressDenyFromRequires(t *testing.T) {
-			Values:   []string{"t"},
-		},
-	})
+	expectedSelector := api.NewESFromMatchRequirements(map[string]string{"any.id": "a"}, nil)
```
What's going on here? The requirement is dropped from the test without explanation. Is this a bug? 🤔

TheBeeZee replied:
That's a very good question. I've put the requirements back and will investigate why this had to be removed.

TheBeeZee replied:
Ok, I think this is related to the discussion about Requirements handling. Right now the requirements are folded only into rules from a single policy, while they should be folded across all policies. This will require a bit of a refactor. I will close this comment thread and keep the one about Requirements open to address this.

joestringer commented on pkg/policy/rule_test.go:
```diff
@@ -301,7 +302,7 @@ func TestMergeL7PolicyIngress(t *testing.T) {
 			{
 				IngressCommonRule: api.IngressCommonRule{
-					FromEndpoints: api.EndpointSelectorSlice{endpointSelectorB},
+					FromEndpoints: []api.EndpointSelector{endpointSelectorB},
```
nit: Seems like unnecessary code churn...?

TheBeeZee replied:
Indeed. Probably a leftover from one of the mass search/replace operations.

joestringer commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1786,7 +1787,7 @@ func PolicyDefaultLocalClusterInspect(ctx context.Context, k8sClient *k8s.Client
 			return nil, fmt.Errorf("Error parsing network policies: %w", err)
 		}
 
-		res.NetworkPolicies[client.ObjectKeyFromObject(&np).String()] = !slices.EqualFunc(rulesAny, rulesLocal, func(a, b *api.Rule) bool { return a.DeepEqual(b) })
+		res.NetworkPolicies[client.ObjectKeyFromObject(&np).String()] = !slices.EqualFunc(rulesAny, rulesLocal, func(a, b *types.PolicyEntry) bool { return a.DeepEqual(b) })
```
Why is the CLI depending on internal implementation details of the cilium-agent?

TheBeeZee replied:
That's a question for the original author of this code. The code parses k8s network policies. Previously the parsing function returned api.Rules, now it returns PolicyEntries:

hhttps://github.com/cilium/cilium/blob/main/cilium-cli/clustermesh/clustermesh.go#L1780

joestringer replied:
I guess there's two queries right here and then a comparison and no reliance on internal details, so this shouldn't cause longer term problems. Still feels wrong to be somehow decoding into unstable internal types in an unrelated component, but I'm not sure I see a better immediate solution. I'll close this thread as resolved.

joestringer commented on pkg/policy/api/selector.go:
```diff
@@ -225,6 +227,7 @@ func NewESFromMatchRequirements(matchLabels map[string]string, reqs []slim_metav
 // become out of sync.
 func (n *EndpointSelector) SyncRequirementsWithLabelSelector() {
 	n.requirements = labelSelectorToRequirements(n.LabelSelector)
+	n.cachedLabelSelectorString = n.LabelSelector.String()
```
Is this a bug or something? Bit surprised to see this in the middle of the big commit.

TheBeeZee replied:
This is mirroring this code:

https://github.com/TheBeeZee/cilium/blob/main/pkg/policy/api/ingress.go#L264

As cachedLabelSelectorString is a private field, I have to update it somewhere. It seems to fit well in SyncRequirementsWithLabelSelector(). If you think otherwise, I can add a separate method for it.

joestringer replied:
Why is this change needed at all? is it related directly to the change of api.Rule -> PolicyEntry? Do tests fail if it's not here? Could this specific change be made independently of the big patch and if yes should we apply it to stable releases to fix a bug?

TheBeeZee replied:
Yes, the change is needed for api.Rule to PolicyEntry migration, specifically for handling Requirements. Previously requirements were handled separately for ingress/egress in policy/pkg/api in functions like this:

https://github.com/TheBeeZee/cilium/blob/main/pkg/policy/api/ingress.go#L245

Because this function updates the MatchExpressions, the cached label selector string also has to be updated. This functionality has now moved into the WithRequirements method in pkg/policy/types/policyentry.go. This method does not have access to internal fields of EndpointSelector. Updating the cached label selector string in SyncREquirementsWithLabelSelector seems to be a natural fit and honestly I'm not sure why it wasn't done this way before.

Unfortunately I forgot to copy the associated comment into the new location, which I have now fixed.

I have also added an extra commit that removes all the dead code in pkg/policy/api related to Requirements field.

Please let me know what you think.

joestringer replied:
Looks good to me. Technically this means that some of the `ToServices` handling code will now cache the label selector string when it did not do so previously. However it seems like that is only used temporarily during initial translation of the services policy rules, and that populated field should not be subsequently used at all. So it's neither a bug in that path nor would I expect the change to introduce a meaningful behavioral difference.

joestringer commented on pkg/policy/types/policyentry.go:
```diff
@@ -0,0 +1,165 @@
+type PolicyEntries []*PolicyEntry
+
+// PeerSelector is a generic representation of an endpoint selector.
+type PeerSelector interface {
+	IsPeerSelector()
+}
```
nit: Still thinking that we can simplify all the type dereference stuff by adding `GetAsEndpointSelectors()` and `DeepEqual()` to this interface. Improvement but not blocking feedback.

TheBeeZee replied:
How would GetAsEndpointSelectors() help? The concrete types that match this interface (EndpointSelector, CIDR, etc.) do not implement it, only the slice versions do.

Are you saying we should add GetAsEndpointSelectors to all concrete types?

joestringer replied:
Yeah just adding to the concrete types.

TheBeeZee replied:
I looked into doing this, but it is not a trivial change. A lot of the current logic for GetAsEndpointSelectors only makes sense in the context of a slice and does not fit well onto a single (concrete) type like EndpointSelector, CIDR or CIDRRule.

I'd like to defer this to a future PR, especially as you seem to think this is a nit.

joestringer replied:
Yeah that's fine, I don't think it's reasonable to increase the scope of this PR further. This is an idea and I think it will improve maintainability of the code, but we may have other lower hanging fruit so I don't want to over-emphasize this feedback.

joestringer reviewed (commented):
There's a couple of minor things that seem out of place but otherwise it seems like this is almost ready to merge.

TheBeeZee committed:
Remove dead policy rule API code dealing with requirements

Signed-off-by: Blaz Zupan <blaz@google.com>

joestringer reviewed (approved):
The changes look good. I wonder whether we may be missing some optimizations with the `aggregatedSelectors` stuff being removed, but it is definitely a cleanup. This should also address a long-running tech debt issue https://github.com/cilium/cilium/issues/8353 🎉 

Last blocker is whether this PR introduces failures in the testsuite. Let's see how the testing results come out.

TheBeeZee committed:
fix: Make PerSelectorPolicy comparison order-insensitive

The recent policy engine refactor introduced non-deterministic ordering of L7Rules within the PerSelectorPolicy struct. This causes test flakiness because the `Equal` method's use of `DeepEqual` is order-sensitive when comparing slices.

To stabilize the comparison, this commit uses `deepequal-gen` annotations to generate a custom `DeepEqual` method that compares L7Rules slices without regard to order. New slice types were introduced for this purpose, as the annotation cannot be applied directly to struct fields.

**Scope:** The Kafka slice is intentionally excluded from this change. Kafka rules are deprecated, and modifying them would require changes to an external package. No test flakiness has been observed for these rules.

Signed-off-by: Blaz Zupan <blaz@google.com>

viktor-kurchenko reviewed (approved):
Thanks!
